prov/gni/src/gnix_rma.c

```diff
@@ -0,0 +1,245 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "gnix.h"
+#include "gnix_nic.h"
+#include "gnix_vc.h"
+#include "gnix_ep.h"
+#include "gnix_mr.h"
+#include "gnix_cm_nic.h"
+#include "gnix_mbox_allocator.h"
+#include <gni_pub.h>
+
+static int __gnix_rma_fab_req_complete(void *arg)
+{
+	struct gnix_fab_req *req = (struct gnix_fab_req *)arg;
+	struct gnix_fid_ep *ep = req->gnix_ep;
+	int rc;
+
+	/* more transaction needed for request? */
+
+	/* write completions */
+	if (ep->send_cq && (!ep->send_selective_completion ||
+			    (req->flags & FI_COMPLETION))) {
+		rc = _gnix_cq_add_event(ep->send_cq, req->user_context,
+					req->flags, req->len,
+					(void *)req->loc_addr,
+					req->data, req->tag);
+		if (rc) {
+			GNIX_WARN(FI_LOG_CQ,
+				  "_gnix_cq_add_event() failed: %d\n", rc);
+		}
+	}
+
+	_fr_free(ep, req);
+
+	return FI_SUCCESS;
+}
+
+static int __gnix_rma_txd_complete(void *arg)
+{
+	struct gnix_tx_descriptor *txd = (struct gnix_tx_descriptor *)arg;
+
+	/* Progress fabric operation in the fab_req completer.  Can we call
+	 * fab_req->completer_func directly from __gnix_tx_progress? */
+	return txd->desc.req->completer_func(txd->desc.req->completer_data);
+}
+
+static gni_post_type_t __gnix_fr_post_type(int fr_type)
+{
+	switch (fr_type) {
+	case GNIX_FAB_RQ_RDMA_WRITE:
+	case GNIX_FAB_RQ_RDMA_WRITE_IMM_DATA:
+		return GNI_POST_RDMA_PUT;
+	case GNIX_FAB_RQ_RDMA_READ:
+		return GNI_POST_RDMA_GET;
+	default:
+		break;
+	}
+
+	GNIX_WARN(FI_LOG_EP_DATA, "Unsupported post type: %d", fr_type);
+	return -FI_ENOSYS;
+}
+
+int _gnix_post_req(struct gnix_fab_req *fab_req)
+{
+	struct gnix_fid_ep *ep = fab_req->gnix_ep;
+	struct gnix_nic *nic = ep->nic;
+	struct gnix_fid_mem_desc *md;
+	struct gnix_tx_descriptor *txd;
+	gni_mem_handle_t mdh;
+	gni_return_t status;
+	int rc;
+
+	rc = _gnix_nic_tx_alloc(nic, &txd);
+	if (rc) {
+		GNIX_INFO(FI_LOG_EP_DATA, "_gnix_nic_tx_alloc() failed: %d\n", rc);
+		return -FI_ENOSPC;
+	}
+
+	txd->desc.completer_func = __gnix_rma_txd_complete;
+	txd->desc.req = fab_req;
+
+	_gnix_convert_key_to_mhdl((gnix_mr_key_t *)&fab_req->rem_mr_key, &mdh);
+	md = (struct gnix_fid_mem_desc *)fab_req->loc_md;
+
+	//txd->desc.gni_desc.post_id = (uint64_t)fab_req; /* unused */
+	txd->desc.gni_desc.type = __gnix_fr_post_type(fab_req->type);
+	txd->desc.gni_desc.cq_mode = GNI_CQMODE_GLOBAL_EVENT; /* check flags */
+	txd->desc.gni_desc.dlvr_mode = GNI_DLVMODE_PERFORMANCE; /* check flags */
+	txd->desc.gni_desc.local_addr = (uint64_t)fab_req->loc_addr;
+	txd->desc.gni_desc.local_mem_hndl = md->mem_hndl;
+	txd->desc.gni_desc.remote_addr = (uint64_t)fab_req->rem_addr;
+	txd->desc.gni_desc.remote_mem_hndl = mdh;
+	txd->desc.gni_desc.length = fab_req->len;
+	txd->desc.gni_desc.rdma_mode = 0; /* check flags */
+	txd->desc.gni_desc.src_cq_hndl = fab_req->gnix_ep->nic->tx_cq; /* check flags */
+
+	{
```

Comment:
This comes down to style, but I don't like using internal scoping to achieve things that should belong in an static/inline function. Could you push this out to another function?